tasmota/tasmota_xsns_sensor/xsns_33_qmc5883l.ino

```diff
@@ -0,0 +1,385 @@
+#define QMC5883L_X_LSB                  0x00
+#define QMC5883L_X_MSB                  0x01
+#define QMC5883L_Y_LSB                  0x02
+#define QMC5883L_Y_MSB                  0x03
+#define QMC5883L_Z_LSB                  0x04
+#define QMC5883L_Z_MSB                  0x05
+#define QMC5883L_STATUS                 0x06
+#define QMC5883L_TEMP_LSB               0x07
+#define QMC5883L_TEMP_MSB               0x08
+#define QMC5883L_CONFIG                 0x09
+#define QMC5883L_CONFIG2                0x0a
+#define QMC5883L_RESET                  0x0b
+#define QMC5883L_RESERVED               0x0c
+#define QMC5883L_CHIP_ID                0x0d
+
+/* Bit values for the STATUS register */
+#define QMC5883L_STATUS_DRDY            1
+#define QMC5883L_STATUS_OVL             2
+#define QMC5883L_STATUS_DOR             4
+
+/* Oversampling values for the CONFIG register */
+#define QMC5883L_CONFIG_OS512           0b00000000
+#define QMC5883L_CONFIG_OS256           0b01000000
+#define QMC5883L_CONFIG_OS128           0b10000000
+#define QMC5883L_CONFIG_OS64            0b11000000
+
+/* Range values for the CONFIG register */
+#define QMC5883L_CONFIG_2GAUSS          0b00000000
+#define QMC5883L_CONFIG_8GAUSS          0b00010000
+
+/* Rate values for the CONFIG register */
+#define QMC5883L_CONFIG_10HZ            0b00000000
+#define QMC5883L_CONFIG_50HZ            0b00000100
+#define QMC5883L_CONFIG_100HZ           0b00001000
+#define QMC5883L_CONFIG_200HZ           0b00001100
+
+/* Mode values for the CONFIG register */
+#define QMC5883L_CONFIG_STANDBY         0b00000000
+#define QMC5883L_CONFIG_CONT            0b00000001
+
+/* Mode values for the CONFIG2 register */
+#define QMC5883L_CONFIG2_RESET          0b10000000
+
+/* Apparently M_PI isn't available in all environments. */
+#ifndef M_PI
+#define M_PI 3.14159265358979323846264338327950288
+#endif
+
+// data field
+struct {
+int16_t   MX = 0, MY = 0, MZ = 0, HG = 0;
+int16_t   xhigh = 0, yhigh = 0, xlow = 0, ylow = 0;
+int16_t   temp = 0;
+uint16_t  scalar = 0;
+uint8_t   status;
+bool      ready = false;
+uint8_t   i2c_address = QMC5883L_ADDR;
+} QMC5883L;
+
+// Initialise the device
+void QMC5883L_Init()
+{
+  if (!I2cSetDevice(QMC5883L.i2c_address)) { return; }
+  I2cSetActiveFound(QMC5883L.i2c_address, "QMC5883L");
+  // reset QMC5883L
+  if (I2cWrite8(QMC5883L.i2c_address, QMC5883L_CONFIG2, QMC5883L_CONFIG2_RESET) == false) return; // Software Reset
+  // writeRegister(QMC5883L_CONFIG2,QMC5883L_CONFIG2_RESET);  // Software Reset
+  //writeRegister(QMC5883L_RESET, 0x01);
+  if (I2cWrite8(QMC5883L.i2c_address, QMC5883L_RESET, 0x01) == false) return;
+  // write config
+  //writeRegister(QMC5883L_CONFIG, QMC5883L_CONFIG_OS256 | QMC5883L_CONFIG_8GAUSS | QMC5883L_CONFIG_10HZ | QMC5883L_CONFIG_CONT);
+  if (I2cWrite8(QMC5883L.i2c_address, QMC5883L_CONFIG, QMC5883L_CONFIG_OS256 | QMC5883L_CONFIG_8GAUSS | QMC5883L_CONFIG_10HZ | QMC5883L_CONFIG_CONT) == false) return;
+  QMC5883L.ready = true;
+}
```

Comment:
Don't use the i2c_address from QMC5883L as it wouldn't have been initialized but use directly the constant `QMC5883L_ADDR`

At the end of the function, if the init is ok, instead of
```
QMC5883L.ready = true;
```
use
```
QMC5883L = calloc(1, sizeof(struct QMC5883L_s));
```
This will initialize all members of the structure to 0 (calloc does that but malloc doesn't)
Now you don't need anymore the `ready` field as you can replace every `if (QMC5883L.ready)`  with `if (QMC5883L)`